File: CCIT2020/training/see_you_in_the_city_center-BRUTE/python/brute.py
#!/usr/bin/python3

import string
import itertools
import base64
import sys
import threading


def status():
    global counter
    print(str(counter / 10) + " key/s")
    t = threading.Timer(10.0, status)
    counter = 0
    t.start()

def encrypt(clear, key):
    enc = []
    for i in range(len(clear)):
        key_c = key[i % len(key)]
        enc_c = chr((ord(clear[i]) + ord(key_c)) % 128)
        enc.append(enc_c)
    return str(base64.urlsafe_b64encode("".join(enc).encode('ascii')), 'ascii')

try_size = 8
key_size = 8

counter = 0
t = threading.Timer(10.0, status)

iter_over = itertools.product(string.ascii_lowercase, repeat = try_size)

m = "See you later in the city center"
c = "QSldSTQ7HkpIJj9cQBY3VUhbQ01HXD9VRBVYSkE6UWRQS0NHRVE3VUQrTDE="


t.start()
for key in iter_over:
    counter += 1
    rev_key = "".join(key)[::-1]
    rev_key += "a" * (key_size - try_size)
    k1 = rev_key[0:4]
    k2 = rev_key[4:8]
    d = encrypt(m, k1)
    tmp_c = encrypt(d, k2)
    print(k1 + k2 + ' - ' + d)
    print(k1 + k2 + ' - ' + tmp_c)
    if (tmp_c == c):
        print("-" * 100)
        print(s1 + s2)
        sys.exit(0)

File: CCIT2020/training/see_you_in_the_city_center-BRUTE/c/brute.c
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

#include "base64.h"
#include "b64.h"

int counter = 0;

void timeout(int sig)
{
	char buf[100];
	sprintf(buf, "%d key/s", counter / 10);
	puts(buf);
	counter = 0;
	alarm(10);
}

char * encrypt(char* clear, int clear_len, char* key, int key_len)
{
	char * enc, * ret;
	int i;
	enc = (char *) malloc(clear_len);
	for (i = 0; i < clear_len; i++){
		enc[i] = (clear[i] + key[i % key_len]) % 128;
	}
	return enc;
}

void print_bin(char * s, int l)
{
	int i;
	for (i = 0; i < l; i++)
		putchar(s[i]);
	putchar('\n');
}

int main(int argc, char ** argv)
{
	char * msg = "See you later in the city center"; //lenght: 32
	int msg_len = strlen(msg);
	char * enc = "QSldSTQ7HkpIJj9cQBY3VUhbQ01HXD9VRBVYSkE6UWRQS0NHRVE3VUQrTDE=";
	int enc_len = strlen(enc);
	char * enc_bin;
	size_t enc_bin_len;
	char *stp1, *stp2, *stp3, *stp4;
	char * key_init = "aaaaaaaa"; // lenght: 8
	char * key;
	int i = 0, j;
	size_t stp1_len, stp2_len, stp3_len, stp4_len;
	key = malloc(9);
	strcpy(key, key_init);
	signal(SIGALRM, timeout);
	if (argc == 2){
		key[7] = argv[1][0];
	}
	enc_bin = b64_decode_ex(enc, enc_len, &enc_bin_len);
	/* debugging
	char * enc_again;
	size_t enc_again_len;
	enc_again = base64_encode(enc_bin, enc_bin_len, &enc_again_len);
	print_bin(enc_again, enc_again_len);
	*/
	alarm(10);
	while(1){
		counter++;
		stp1 = encrypt(msg, msg_len, key, 4);
		stp2 = base64_encode(stp1, msg_len, &stp2_len);
		free(stp1);
		stp3 = encrypt(stp2, stp2_len, key + 4, 4);
		free(stp2);
		/* debug 
		stp4 = base64_encode(stp3, stp2_len, &stp4_len);
		printf("%s - %s\n", key, stp4);
		*/
		if (!memcmp(stp3, enc_bin,  enc_bin_len)){
			stp4 = base64_encode(stp3, stp2_len, &stp4_len);
			printf("found\n%s - %s\n", key, stp4);
			exit(0);
		}
		free(stp3);
		for (i = 0; i < 7; i++){
			if (key[i] < 'z'){
				for (j = i - 1; j >= 0; j--){
					key[j] = 'a';
				}
				key[i]++;
				break;
			}
		}
		if (i == 7){
			exit(1);
		}
	}
}
